proj/hps_accel/gateware/sp_mem.py

```diff
@@ -0,0 +1,170 @@
+        return unsigned((self.depth - 1).bit_length())
+
+    def data_shape(self) -> Shape:
+        """nMigen shape describing a data word."""
+        return unsigned(self.width)
+
+    def write_stream_payload_type(self) -> Layout:
+        """Gets the payload type for a write stream"""
+        return Layout([
+            ("addr", self.addr_shape(), DIR_FANOUT),
+            ("data", self.data_shape(), DIR_FANOUT),
+        ])
+
+    def make_write_sink(self):
+        return Sink(self.write_stream_payload_type())
+
+    def make_write_source(self):
+        return Source(self.write_stream_payload_type())
+
+    def make_read_addr_sink(self):
+        return Sink(self.addr_shape())
+
+    def make_read_addr_source(self):
+        return Source(self.addr_shape())
+
+    def make_read_data_sink(self):
+        return Sink(self.data_shape())
+
+    def make_read_data_source(self):
+        return Source(self.data_shape())
+
+
+class SinglePortMemory(SimpleElaboratable):
+    """A single port memory with stream interfaces.
+
+    Data is written via the memory's write_sink, which has a payload of
+    the data to write and the address to write it to.
+
+    Data is read via the memory's read_addr_sink and read_data_source.
+    For each address sent to the sink, one word is returned on the
+    source.
+
+    Writes have priority over reads. The read addr sink will not become
+    ready if there is unread data waiting to be transferred.
+
+    Parameters
+    ----------
+    params: MemoryParameters
+      Describes the memory.
+
+    Attributes
+    ----------
+
+    write_sink: Sink[addr, data]
+      For each packet sent on this sink, one word will be written to
+      the memory.
+
+    read_addr_sink: Sink[addr]
+      For each packet sent on this sink, one word will be read.
+
+    read_data_source: Source[data]
+
+    TODO: This same interface would work for pseudo-dual port memory.
+          Refactor.
+    """
+
+    def __init__(self, params: MemoryParameters):
+        self.params = params
+        self.write_sink = params.make_write_sink()
+        self.read_addr_sink = params.make_read_addr_sink()
+        self.read_data_source = params.make_read_data_source()
+
+    def elab(self, m: Module):
+        memory = Memory(depth=self.params.depth, width=self.params.width)
+        m.submodules["read_port"] = read_port = memory.read_port(
+            transparent=False)
+        m.submodules["write_port"] = write_port = memory.write_port()
+
+        # Write sink always ready to write - set address and data directly
+        # to memory.
+        m.d.comb += self.write_sink.ready.eq(1)
+        with m.If(self.write_sink.is_transferring()):
+            m.d.comb += [
+                write_port.en.eq(1),
+                write_port.addr.eq(self.write_sink.payload.addr),
+                write_port.data.eq(self.write_sink.payload.data),
+            ]
+
+        m.d.comb += read_port.en.eq(~write_port.en)
```

Comment:
Let me try to check my understanding based on what we discussed this morning...

Here you are using nmigen Memory, which models a "pseudo dual port memory" (meaning it has one port which can only read, and another port which can only write). You are using it to implement a single port memory, by effectively just turning on only one of either the read port or the write port at any given time, in this statement.

Are some of the synthesis tools smart enough to infer from this that the memory needn't actually be dual ported? But yosys isn't, so it will end up trying to construct a dual-port memory where half the bandwidth is wasted?

On Arty that waste doesn't matter because we have heaps of spare resources. But on Crosslink-NX we will need to avoid nmigen Memory and instead configure the `SP16K` EBRAM primitive. That's important because `SP16K` lets us do up to 36 bit width from each EBRAM, meaning we can do a 128 bit memory using 4 EBRAMs instead of 8.

Am I grokking it?

Reply:
Yes, that's all correct.